From 8a74962dfb8a97da57c1efc82a2d1f1db7c8fc6b Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:31 +0800
Subject: [PATCH] arm64/mpam: update group flags only when enable sucsses
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 8a74962dfb8a97da57c1efc82a2d1f1db7c8fc6b


hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

The ctrlcon flag should be set only when we sucessful to
enable ctrlmon, or the flag is mismatch with ctrlmon status.

ctrlmon is indicate a group which could moniter the usage
of L3/Memory bandwidth.

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/mpam.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/mpam.c b/arch/arm64/kernel/mpam.c
index a9dab7eb7b77..dae2303b2449 100644
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -1013,9 +1013,10 @@ static ssize_t resctrl_group_ctrlmon_write(struct kernfs_open_file *of,
 		rdtgrp->flags &= ~RDT_CTRLMON;
 		resctrl_ctrlmon_disable(rdtgrp->mon.mon_data_kn, rdtgrp);
 	} else if (!(rdtgrp->flags & RDT_CTRLMON) && ctrlmon) {
-		rdtgrp->flags |= RDT_CTRLMON;
-		resctrl_ctrlmon_enable(rdtgrp->kn, rdtgrp,
-				       &rdtgrp->mon.mon_data_kn);
+		ret = resctrl_ctrlmon_enable(rdtgrp->kn, rdtgrp,
+					     &rdtgrp->mon.mon_data_kn);
+		if (!ret)
+			rdtgrp->flags |= RDT_CTRLMON;
 	} else {
 		ret = -ENOENT;
 	}
-- 
2.26.2

